src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesVsPatterns.cs

```diff
@@ -2856,5 +2856,400 @@ public bool Test()
+                );
+        }
+
+        [Fact, WorkItem(65976, "https://github.com/dotnet/roslyn/issues/65976")]
+        public void LoopWithPatternDeclaration_Tuple()
+        {
+            var source = """
+#nullable enable
+using System.Collections.Generic;
+
+class C
+{
+    void M()
+    {
+        string? s = "";
+
+        for (var x = 0; x < 10; x++)
+        {
+            var a = Infer(s);
+            if ((a[0], 1) is (var z, var z2))
+            {
+                z.ToString();
+            }
+
+            s = null;
+        }
+    }
+
+    List<T> Infer<T>(T t) => new() { t };
+}
+""";
+            var comp = CreateCompilation(source);
+            comp.VerifyDiagnostics(
+                // (15,17): warning CS8602: Dereference of a possibly null reference.
+                //                 z.ToString();
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "z").WithLocation(15, 17)
+                );
+        }
+
+        [Fact, WorkItem(65976, "https://github.com/dotnet/roslyn/issues/65976")]
+        public void LoopWithPatternDeclaration_ListPattern()
+        {
+            var source = """
+#nullable enable
+using System.Collections.Generic;
+
+string? s = "";
+
+for (var x = 0; x < 10; x++)
+{
+    var a = Infer(s);
+    if (a is [var z])
+    {
+        z.ToString();
+    }
+
+    s = null;
+}
+
+List<T> Infer<T>(T t) => new() { t };
+""";
+            var comp = CreateCompilation(source, targetFramework: TargetFramework.Net70);
+            comp.VerifyDiagnostics(
+                // (11,9): warning CS8602: Dereference of a possibly null reference.
+                //         z.ToString();
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "z").WithLocation(11, 9)
+                );
+        }
+
+        [Fact, WorkItem(65976, "https://github.com/dotnet/roslyn/issues/65976")]
+        public void LoopWithPatternDeclaration_ListPattern_Inline()
+        {
+            var source = """
+#nullable enable
+using System.Collections.Generic;
+
+string? s = "";
+
+for (var x = 0; x < 10; x++)
+{
+    if (Infer(s) is [var z])
+    {
+        z.ToString();
+    }
+
+    s = null;
+}
+
+List<T> Infer<T>(T t) => new() { t };
+""";
+            var comp = CreateCompilation(source, targetFramework: TargetFramework.Net70);
+            comp.VerifyDiagnostics(
+                // (10,9): warning CS8602: Dereference of a possibly null reference.
+                //         z.ToString();
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "z").WithLocation(10, 9)
+                );
+        }
+
+        [Fact, WorkItem(65976, "https://github.com/dotnet/roslyn/issues/65976")]
+        public void LoopWithPatternDeclaration_SlicePattern()
+        {
+            var source = """
+#nullable enable
+
+string? s = "";
+
+for (var x = 0; x < 10; x++)
+{
+    var a = Infer(s);
+    if (a is [_, .. var z, _])
+    {
+        z.ToString();
```

Comment:
>z [](http://example.com/codeflow?start=8&length=1)

I think it would be good to test scenario where slice returns ```Collection<T>``` and here we access its element. I am assuming a warning is expected for that scenario.